QoL/Modules/NPCSellAll.cs

```diff
@@ -109,5 +123,65 @@ private static void SellEggs()
 
             pd.SetInt(nameof(pd.rancidEggs), 0);
         }
+
+        private static bool _spokeOnRight;
+
+        private static void NailsmithBuy(Scene scene)
+        {
+            GameObject nailsmith = scene.GetRootGameObjects().FirstOrDefault(obj => obj.name == "Nailsmith");
+
+            PlayMakerFSM nailsmithConvo = nailsmith.LocateMyFSM("Conversation Control");
+            FsmState boxUp = nailsmithConvo.GetState("Box Up");
+            boxUp.AddTransition("BOUGHT ALL", "Box Up 4");
+            boxUp.InsertMethod(0, () =>
+            {
+                if (_spokeOnRight && BuyNailUpgrades())
+                {
+                    nailsmithConvo.SendEvent("BOUGHT ALL");
+                }
+            });
+
+            PlayMakerFSM nailsmithRegion = nailsmith.LocateMyFSM("npc_control");
+            nailsmithRegion.GetState("Move Hero Left").InsertMethod(0, () =>
+            {
+                float heroMin = HeroController.instance.GetComponent<Collider2D>().bounds.min.x;
+                // True if the knight is completely to the right of the Nailsmith's dream dialogue hitbox
+                _spokeOnRight = 18.5f < heroMin;
+            });
+        }
+
+        private static bool BuyNailUpgrades()
+        {
+            var pd = PlayerData.instance;
+
+            int current = pd.GetInt(nameof(pd.nailSmithUpgrades));
+            if (current > 3) return false;
+
+            int bought = 0;
+
+            for (int i = current; 
```

Comment:
Use LINQ here, maybe `foreach (var (ore, geo) in NailUpgradeCosts.Skip(current).TakeWhile(x => x.ore <= pd.GetInt(nameof(pd.ore)) && x.geo <=pd.GetInt(nameof(pd.geo)))`

Reply:
This relies on TakeWhile being evaluated lazily (because pd.geo and pd.ore change throughout the evaluation) - would that be an implementation detail of TakeWhile or is it safe to use?

Reply:
Nah, LINQ is explicitly Lazy